src/goto-programs/goto_trace.h

```diff
@@ -138,6 +136,22 @@ class goto_trace_stept
     full_lhs_value.make_nil();
     cond_expr.make_nil();
   }
+
+  goto_programt::const_targett get_pc() const { return pc; }
+
+  /// We don't have a way of enforcing this, but \p pc_ must *not* be a
+  /// singular iterator!
+  void set_pc(goto_programt::const_targett pc_)
+  {
+    pc = std::move(pc_);
+    is_pc_set = true;
+  }
+
+  bool get_is_pc_set() const { return is_pc_set; }
```

Comment:
It forces users to think about an initializer and and, by making it private, removes the capability of changing it later on. This achieves, quoting your earlier message, "The solution is to make the pc field of goto_trace_stept private, and to record when it is set to a (hopefully) non-singular iterator." -- without having to introduce an extra Boolean variable to do so.